text/0113-secure-script-container.md

```diff
@@ -0,0 +1,181 @@
+Many LWC applications apply the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native shadow behavior while still allowing global styles to cascade into the shadow trees. Synthetic shadow currently [patches](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/src/polyfills/document-shadow) the `document.querySelector` and `document.querySelectorAll` APIs which is what prevents developers from penetrating the shadow DOM boundaries either in their LWC components or any other scripts added to the document. We only want to solve for the latter scenario.
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components, we are currently proposing the following characteristics for the secure-script design:
+
+* Create a secure JS sandbox which doesn’t allow its object graph to be extended using [`Object.preventExtensions`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/preventExtensions)
+  * The sandbox is created using a same-domain iframe provided by Locker v.Next
+  * This has to be done before the synthetic shadow is imported
+
+* Run or include the third party library script inside this sandbox
+  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow
+* Package the above functionality in a web component which semantically matches the `<script>` tag.
+
+
+### Native Shadow
+
+For applications which are using the native shadow, the underlying implementation will be different. There are two approaches we can choose from:
+
+1. Virtualize every shadow DOM tree in the outer window into a single DOM structure which this container would see as the default DOM.
+2. Implement a `document.querySelector` API which traverses the shadow DOM naturally.
+
+### Globals
+
+Some scripts define globals which are meant to be used by the rest of the application. Let's take this example of Google Analytics:
+
+```html
+<!-- Google Analytics -->
+<script>
+    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</script>
+<script async src='https://www.google-analytics.com/analytics.js'></script>
+<!-- End Google Analytics -->
+```
+
+If we included the above snippet inside our secure-script container, then `ga` will only be available to the *inner* `window`, i.e. the secure container's window. So `ga` will not be usable by anyone else in the entire application. We can however, allow for globals to be created inside our secure container, and redefined in any additional application sandboxes created by Locker (or the global outer window).
+
+A syntax like the following could be provided:
+
+```html
+<secure-script extra-globals="ga"></secure-script>
```

Comment:
> Just to confirm extra-globals="ga" means that the secure script can set the ga property on the global object. Right?

Yes, more specifically it means that any globals defined by a script in the secure container, are re-set in the outer window.

The secure script container does not have access to globals defined by other scripts. i.e. extra-globals does not define globals to keep in sync between the outer window and the secure script container. Rather it should copy any globals defined to the outer window.

So in your example it would log `undefined` even if `extra-globals` was set. I admit `extra-globals` may imply it's a pass-in.

Reply:
> The secure script container does not have access to globals defined by other scripts

If it's the case, I don't see how the Google Analytics example works. The `analytics.js` does expect `ga` to be available on the `window` in order to run.

Reply:
The secure script would be the first one, as I said above, which defines the globals, like this:

```html
<!-- Google Analytics -->
<secure-script extra-globals="ga">
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-XXXXX-Y', 'auto');
    ga('send', 'pageview');
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>
<!-- End Google Analytics -->
```

Reply:
Shouldn't the analytics scripts be executed as well in a `secure-script` in order to work across shadows transparently?

Reply:
yes, both should be executed under the `<virtual-script>`

Reply:
Yes, the point here was just to illustrate the global being referenced across script containers, both virtual and standard.

Reply:
> The secure script container does not have access to globals defined by other scripts. i.e. extra-globals does not define globals to keep in sync between the outer window and the secure script container. Rather it should copy any globals defined to the outer window.

I think this should be defined in the spec, at least as a note.

> Yes, the point here was just to illustrate the global being referenced across script containers, both virtual and standard.

Does this negate @pmdartus point that due to not being able to access other globals that GA won't work?

Reply:
I think it seems like a good best practice but technically it should negate @pmdartus point since it would work, any globals created will be available in all sandboxes, and yeah I'll add that to the spec.